programs/uxd/src/utils/maths.rs

```diff
@@ -0,0 +1,406 @@
+where
+    T: num_traits::PrimInt + Display,
+{
+    if let Some(res) = arg1.checked_div(&arg2) {
+        Ok(res)
+    } else {
+        msg!("Error: Overflow in {} / {}", arg1, arg2);
+        err!(UxdError::MathOverflow)
+    }
+}
+
+pub fn checked_float_div<T>(arg1: T, arg2: T) -> Result<T>
+where
+    T: num_traits::Float + Display,
+{
+    if arg2 == T::zero() {
+        msg!("Error: Overflow in {} / {}", arg1, arg2);
+        return err!(UxdError::MathOverflow);
+    }
+    let res = arg1 / arg2;
+    if !res.is_finite() {
+        msg!("Error: Overflow in {} / {}", arg1, arg2);
+        err!(UxdError::MathOverflow)
+    } else {
+        Ok(res)
+    }
+}
+
+pub fn checked_ceil_div<T>(arg1: T, arg2: T) -> Result<T>
+where
+    T: num_traits::PrimInt + Display,
+{
+    if arg1 > T::zero() {
+        if arg1 == arg2 && arg2 != T::zero() {
+            return Ok(T::one());
+        }
+        if let Some(res) = (arg1 - T::one()).checked_div(&arg2) {
+            Ok(res + T::one())
+        } else {
+            msg!("Error: Overflow in {} / {}", arg1, arg2);
+            err!(UxdError::MathOverflow)
+        }
+    } else if let Some(res) = arg1.checked_div(&arg2) {
+        Ok(res)
+    } else {
+        msg!("Error: Overflow in {} / {}", arg1, arg2);
+        err!(UxdError::MathOverflow)
+    }
+}
+
+pub fn checked_decimal_div(
+    coefficient1: u64,
+    exponent1: i32,
+    coefficient2: u64,
+    exponent2: i32,
+    target_exponent: i32,
+) -> Result<u64> {
+    if coefficient2 == 0 {
+        msg!("Error: Overflow in {} / {}", coefficient1, coefficient2);
+        return err!(UxdError::MathOverflow);
+    }
+    if coefficient1 == 0 {
+        return Ok(0);
+    }
+    // compute scale factor for the dividend
+    let mut scale_factor = 0;
+    let mut target_power = checked_sub(checked_sub(exponent1, exponent2)?, target_exponent)?;
+    if exponent1 > 0 {
+        scale_factor = checked_add(scale_factor, exponent1)?;
+    }
+    if exponent2 < 0 {
+        scale_factor = checked_sub(scale_factor, exponent2)?;
+        target_power = checked_add(target_power, exponent2)?;
+    }
+    if target_exponent < 0 {
+        scale_factor = checked_sub(scale_factor, target_exponent)?;
+        target_power = checked_add(target_power, target_exponent)?;
+    }
+    let scaled_coeff1 = if scale_factor > 0 {
+        checked_mul(
+            coefficient1 as u128,
+            checked_pow(10u128, scale_factor as usize)?,
+        )?
+    } else {
+        coefficient1 as u128
+    };
+
+    if target_power >= 0 {
+        checked_as_u64(checked_mul(
+            checked_div(scaled_coeff1, coefficient2 as u128)?,
+            checked_pow(10u128, target_power as usize)?,
+        )?)
+    } else {
+        checked_as_u64(checked_div(
+            checked_div(scaled_coeff1, coefficient2 as u128)?,
+            checked_pow(10u128, (-target_power) as usize)?,
+        )?)
+    }
+}
+
+pub fn checked_decimal_ceil_div(
```

Comment:
some of those i dont see us using it a lot, is that a lib imported from somewhere ? or is that functions used for the future ? do we want to write unit tests for some of those ?

Reply:
It's all used by several solana labs projects, audited several times, I kept the whole thing (I think it's better to use that and go there when we need something than rewrite. Would keep as it and update when they do from time to time)